src/main/java/com/google/devtools/build/lib/bazel/repository/starlark/StarlarkBaseExternalContext.java

```diff
@@ -1150,29 +1157,115 @@ protected StarlarkPath getPath(String method, Object path)
       doc = "Reads the content of a file on the filesystem.",
       useStarlarkThread = true,
       parameters = {
-          @Param(
-              name = "path",
-              allowedTypes = {
-                  @ParamType(type = String.class),
-                  @ParamType(type = Label.class),
-                  @ParamType(type = StarlarkPath.class)
-              },
-              doc = "path of the file to read from."),
+        @Param(
+            name = "path",
+            allowedTypes = {
+              @ParamType(type = String.class),
+              @ParamType(type = Label.class),
+              @ParamType(type = StarlarkPath.class)
+            },
+            doc = "path of the file to read from."),
+        @Param(
+            name = "watch",
+            defaultValue = "True",
+            doc = "whether to <a href=\"#watch\">watch</a> the file.")
       })
-  public String readFile(Object path, StarlarkThread thread)
+  public String readFile(Object path, boolean watch, StarlarkThread thread)
       throws RepositoryFunctionException, EvalException, InterruptedException {
     StarlarkPath p = getPath("read()", path);
     WorkspaceRuleEvent w =
         WorkspaceRuleEvent.newReadEvent(
             p.toString(), getIdentifyingStringForLogging(), thread.getCallerLocation());
     env.getListener().post(w);
+    if (watch) {
+      watch(p);
+    }
     try {
       return FileSystemUtils.readContent(p.getPath(), ISO_8859_1);
     } catch (IOException e) {
       throw new RepositoryFunctionException(e, Transience.TRANSIENT);
     }
   }
 
+  @Nullable
+  protected Pair<RepoRecordedInput.File, RootedPath> toRootedPath(Path path) {
+    if (path.startsWith(workingDirectory)) {
+      // The file is under the working directory. Don't watch it, as it would cause a dependency
+      // cycle.
+      return null;
+    }
+    if (path.startsWith(directories.getWorkspace())) {
```

Comment:
Can't these also cause dependency cycles? E.g. if repo `@a` depends on repo `@b` and repo `@b` watches a file in repo `@a`?

Maybe disallow watching files in other repositories by absolute path and then I *think* the `getRootedPathFromLabel()` in `getPathFromLabel()` would at least make the error message a little less confusing?

Reply:
these only cause cycles if there _are_ cycles (unlike the first case, which is an immediate guaranteed cycle). Normally it's perfectly fine if `@a` watches something in `@b`.

> Maybe disallow watching files in other repositories by absolute path and then I _think_ the `getRootedPathFromLabel()` in `getPathFromLabel()` would at least make the error message a little less confusing?

I don't really understand this part. Could you elaborate / rephrase?

Reply:
It looks like your intent is that one would watch files in, say, repository `@a` by saying `rctx.watch("@a//pkg:file")` (even if `pkg/BUILD` does not exist)

But what stops the user from watching the same file by passing in its absolute path by plumbing in the output base by whatever ugly approach they choose?

Not that prohibiting this would make the dependency cycle to away, but hopefully it would yield a nicer error message. Also, maybe add a test case to ensure that Bazel does not crash in this case?

Reply:
> But what stops the user from watching the same file by passing in its absolute path by plumbing in the output base by whatever ugly approach they choose?

nothing prevents it, it's the same as using the label. (See my reply to the other comment)

> Also, maybe add a test case to ensure that Bazel does not crash in this case?

I'll add a test case for the cycle.